Update to electron-forge 6 & simplify build pipeline/project structure (#418)
adopted-ember-addons/ember-electron: tests/dummy/app/templates/docs/guides/upgrading.md

# Upgrading from ember-electron 2.x

While upgrading from 2.x to 3.x should not be very time-consuming in most cases, quite a bit has changed, so you should follow these steps in order to be successful. Starting from a clean tree (so you can run a clean `git diff` afterwards to see what changed), the recommended steps are:

## Upgrade ember electron

```sh
yarn-or-npm remove ember-electron
ember install ember-electron
```

When prompted to overwrite `testem-electron.js`, choose `yes`.

## Fix electron version

Your `electron-app` project will initially have the most recent release of Electron installed. You were probably running an older version since it was very difficult to run newer than Electron 4.x with `ember-electron` 2.x. So if you don't want to deal with upgrading Electron in the middle of upgrading `ember-electron`, you should change your `electron-app/package.json`'s `electron` version to match the version of `electron-prebuilt-compile` (which would determine the Electron version you were running) that is in your root `package.json`. Don't forget to run `yarn`/`npm install` from `electron-app` afterwards!

> The good news is that now upgrading Electron just involves updating the version installed in the `electron-app` package -- no more waiting for `electron-prebuilt-compile` to release an update matching the latest Electron version!

Reviewer: ```suggestion
The good news is that now upgrading Electron only involves updating the version installed in the `electron-app` package -- no more waiting for `electron-prebuilt-compile` to release an update matching the latest Electron version!
```

Author: (same comment here as above)

Reviewer: I'm not the one who started the war on "just", I know it gets flagged by a lot of tools as being discriminatory or something though.

Author: Like is it an ESL thing? Can you link to any articles or anything?

To be blunt, I'm not inclined to follow advice that seems wrong to me if I don't have any understanding of the reasoning behind it. I'm certainly not going to die on this hill if you feel really strongly about it, but even after a little googling, I'm unable to get any sense of why `only` would be preferable to `just` here.

Reviewer: I think the argument is it has a negative connotation like "just do it, it's easy", "just make it work" etc. This case doesn't necessarily apply. I don't particularly care that much, only following what I have heard. There was a talk at EmberConf about it, and it sounds like whatever tool they use only looks for how many times you said "just", "guys", etc anything that could be referring to gender or implying that something is easy, when it is not easy for everyone.

```javascript
protocolServe({
  cwd: join(__dirname || resolve(dirname('')), '..', 'ember'),
  app,
  protocol,
});
```

while `ember-electron` 3.x's contains:

```
protocolServe({
  cwd: join(__dirname || resolve(dirname('')), '..', 'ember-dist'),
  app,
  protocol,
});
```

### test-main.js

Generally you should not need to migrate `test-main.js` at all. It's default content in 2.x was:

```javascript
/* eslint-env node */
require('ember-electron/lib/test-support/test-main');
```

If yours still looks like that, then you don't need to do anything. If you've customized it, then you'll need to migrate the content to `electron-app/tests/index.js`

### resources-* folders

If you don't have any files in any of the `ember-electron/resources` or `ember-electron/resources-*` folders, you can skip this section.

`ember-electron` 2.x managed the `resources-*` folders to allow you to specify platform-specific resources. `electron-forge` does not support this functionality, and it was one of the main factors contributing to the complexity and slowness of `ember-electron` 2.x's build pipeline, so it's been removed in 3.x.

> So if you only have content in `ember-electron/resources/`, you can just copy the folder into `electron-app`. Note that `src/index.js` is now in a subfolder, unlike `main.js` was in 2.x, so if you are accessing resources from your main process using, e.g., `path.join(__dirname, 'resources')`, you'll have to update it to `path.join(__dirname, '..', 'resources')` (or you could put the `resources` folder in the `src/` directory if you're some kind of monster).

Reviewer: ```suggestion
So if you only have content in `ember-electron/resources/`, you can copy the folder into `electron-app`. Note that `src/index.js` is now in a subfolder, unlike `main.js` was in 2.x, so if you are accessing resources from your main process using, e.g., `path.join(__dirname, 'resources')`, you'll have to update it to `path.join(__dirname, '..', 'resources')`. Alternatively, you could put the `resources` folder in the `src/` directory, but it is not recommended.
```

Author: The language here is trying to emphasize that this is all you need to do as far as copying files if you don't have any platform-specific resources, but if you do have platform-specific resources, you need to do some other stuff.

Maybe

> So if you only have content in `ember-electron/resources/`, you can copy the folder into `electron-app` and ignore the other `ember-electron/resources-*` folders.

?

Reviewer: Seems fine to me. 👍
